### PR TITLE
scheduler: merge Reserve and Unreserve plugins

### LINKED ISSUE #90724: Scheduling Framework: Merge Reserve and Unreserve
/kind cleanup

It doesn't make much sense for Reserve plugins to not have an Unreserve counterpart. Having them separate adds configuration toil.
I propose we merge them in a single interface and a single PluginSet in the configuration. However, we can only do the later in a new version of component config.

/sig scheduling

Related #83557 #83726  kubernetes-sigs/scheduler-plugins#4

### PR DESCRIPTION
**Pull request by adtac**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
/kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Previously, separate interfaces were defined for Reserve and Unreserve
plugins. However, in nearly all cases, a plugin that allocates a
resource using Reserve will likely want to register itself for Unreserve
as well in order to free the allocated resource at the end of a failed
scheduling/binding cycle. Having separate plugins for Reserve and
Unreserve also adds unnecessary config toil. To that end, this patch
aims to merge the two plugins into a single interface called a
ReservationPlugin that requires implementing both the Reserve and
Unreserve methods.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/90724

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Unreserve extension point for scheduler plugins is merged into Reserve extension point
```

### CODE REVIEW

**Comment by adtac** [comment]
CLA signed.

File: pkg/scheduler/apis/config/types.go [hunk 1]
```diff
@@ -192,8 +192,8 @@ type Plugins struct {
 	// Score is a list of plugins that should be invoked when ranking nodes that have passed the filtering phase.
 	Score *PluginSet
 
-	// Reserve is a list of plugins invoked when reserving a node to run the pod.
-	Reserve *PluginSet
+	// Reservation is a list of plugins invoked when reserving and unreserving a node to run the pod.
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types.go:195]: I would prefer that we keep the name `Reserve` to avoid increasing the backwards-incompatible changes from v1alpha2 to v1beta1.
I don't think the name Reserve is deceiving. We have `Score`, which has 2 functions: `Score` and `NormalizeScore`.

WDYT @ahg-g @Huang-Wei ?

File: pkg/scheduler/framework/plugins/examples/multipoint/multipoint.go [hunk 2]
```diff
@@ -63,7 +63,24 @@ func (mc CommunicatingPlugin) Reserve(ctx context.Context, state *framework.Cycl
 	return nil
 }
 
-// PreBind is the functions invoked by the framework at "prebind" extension point.
+// Unreserve is the function invoked by the framework at "unreserve" extension
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/examples/multipoint/multipoint.go:66]: Unreserve is the function invoked by the framework when any error happens during "reserve" extension point or later.

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 3]
```diff
@@ -770,26 +768,27 @@ func (f *framework) runPostBindPlugin(ctx context.Context, pl PostBindPlugin, st
...
 
-// RunReservePlugins runs the set of configured reserve plugins. If any of these
-// plugins returns an error, it does not continue running the remaining ones and
-// returns the error. In such case, pod will not be scheduled.
-func (f *framework) RunReservePlugins(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) (status *Status) {
+// RunReserveReservationPlugins runs the Reserve method in the set of
+// configured reservation plugins. If any of these plugins returns an error, it
+// does not continue running the remaining ones and returns the error. In such
+// case, pod will not be scheduled.
+func (f *framework) RunReserveReservationPlugins(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) (status *Status) {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/v1alpha1/framework.go:775]: `RunReservePluginsReserve` if we revert the change about Reservation

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 4]
```diff
@@ -799,18 +798,20 @@ func (f *framework) runReservePlugin(ctx context.Context, pl ReservePlugin, stat
 	return status
 }
 
-// RunUnreservePlugins runs the set of configured unreserve plugins.
-func (f *framework) RunUnreservePlugins(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) {
+// RunUnreserveReservationPlugins runs the set of configured unreserve plugins.
+func (f *framework) RunUnreserveReservationPlugins(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/v1alpha1/framework.go:802]: `RunReservePluginsUnreserve`

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 5]
```diff
@@ -324,17 +324,20 @@ type ScorePlugin interface {
...
+// used to be called 'assume' in the original scheduler. These plugins should
+// return only Success or Error in Status.code. However, the scheduler accepts
+// other valid codes as well. Anything other than Success will lead to
+// rejection of the pod.
+type ReservationPlugin interface {
 	Plugin
 	// Reserve is called by the scheduling framework when the scheduler cache is
 	// updated.
 	Reserve(ctx context.Context, state *CycleState, p *v1.Pod, nodeName string) *Status
+	// Unreserve is called by the scheduling framework when the scheduler cache is
```
Comment (alculquicondor) [inline pkg/scheduler/framework/v1alpha1/interface.go:338]: `Unreserve is called by the scheduling framework when a reserved pod was rejected or an error occurred during reservation of other plugins or in a later phase. Unreserve must be idempotent.`

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -522,8 +522,8 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 	assumedPodInfo := podInfo.DeepCopy()
 	assumedPod := assumedPodInfo.Pod
 
-	// Run "reserve" plugins.
-	if sts := prof.RunReservePlugins(schedulingCycleCtx, state, assumedPod, scheduleResult.SuggestedHost); !sts.IsSuccess() {
+	// Run the Reserve method of reservation plugins.
+	if sts := prof.RunReserveReservationPlugins(schedulingCycleCtx, state, assumedPod, scheduleResult.SuggestedHost); !sts.IsSuccess() {
 		sched.recordSchedulingFailure(prof, assumedPodInfo, sts.AsError(), SchedulerError, "")
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:527]: Here we should call Unreserve #83557

We will need a test for it.

File: pkg/scheduler/apis/config/types.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/apis/config/types.go:195]: keeping the name Reserve sgtm.

File: pkg/scheduler/scheduler.go (hunk 6 above)
Comment (adtac) [reply pkg/scheduler/scheduler.go:527]: I'd like to implement the merge first and then handle the case of failing reserves later separately to keep the PR small, atomic, and contained to one issue.

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 5 above)
Comment (adtac) [reply pkg/scheduler/framework/v1alpha1/interface.go:338]: Can you please elaborate why plugin authors must guarantee the idempotency of Unreserve operations? AFAIK there is no code path that will call the Unreserve method of a plugin more than once. In that case, whether or not a plugin's Unreserve is idempotent is immaterial. Most unreserves will likely be idempotent, but not all have to be (and it doesn't matter to us whether or not they are).

File: pkg/scheduler/apis/config/types.go (hunk 1 above)
Comment (adtac) [reply pkg/scheduler/apis/config/types.go]: Done.

File: pkg/scheduler/scheduler.go (hunk 6 above)
Comment (mrkm4ntr) [reply pkg/scheduler/scheduler.go:527]: Could I recreate https://github.com/kubernetes/kubernetes/pull/83557 after this PR will be merged?

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/v1alpha1/interface.go]: It will depend on how you choose to implement #83557

A simple approach I was thinking of is just calling RunReservePluginsUnreserve. But to do that, Unreserve calls need to be idem-potent.

On the other hand, you could just call Unreserve directly in RunReservePluginsReserve, where you know which plugins succeded to run. But then there is the question: should we call Unreserve for the plugin that just failed?

In any case, we can leave this discussion for the follow up PR.

File: pkg/scheduler/scheduler.go (hunk 6 above)
Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:527]: @adtac could take it, since he is already modifying the code here. Unless you @mrkm4ntr  have some specific need?

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 5 above)
Comment (adtac) [reply pkg/scheduler/framework/v1alpha1/interface.go]: Agreed, let's leave the discussion of idempotency to the other PR. (FWIW, since we know the index of the first reserve plugin that failed, we should be able to call Unreserve on just the other plugins that were previously reserved; whether or not the Unreserve should be called on the plugin that failed is a good question, though.)

File: pkg/scheduler/scheduler.go (hunk 6 above)
Comment (adtac) [reply pkg/scheduler/scheduler.go:527]: I'd like to see this feature to completion, but if you'd strongly like to work on that feature after this PR, I don't mind :)

File: test/integration/scheduler/framework_test.go [hunk 7]
```diff
@@ -958,24 +944,26 @@ func TestPrebindPlugin(t *testing.T) {
 	}
 }
 
-// TestUnreservePlugin tests invocation of un-reserve plugin
-func TestUnreservePlugin(t *testing.T) {
-	// TODO: register more plugin which would trigger un-reserve plugin
+// TestUnreserveReservePlugin tests invocation of the unreserve operation
+// in reserve plugins.
+func TestUnreserveReservePlugin(t *testing.T) {
+	// TODO: register more plugin which would trigger unreserve in reserve
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:950]: Put in your username or an issue: `TODO(adtac)`, `TODO(#12345)`

File: test/integration/scheduler/framework_test.go [hunk 8]
```diff
@@ -958,24 +944,26 @@ func TestPrebindPlugin(t *testing.T) {
 	}
 }
 
-// TestUnreservePlugin tests invocation of un-reserve plugin
-func TestUnreservePlugin(t *testing.T) {
-	// TODO: register more plugin which would trigger un-reserve plugin
+// TestUnreserveReservePlugin tests invocation of the unreserve operation
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:947]: nit: `when PreBind fails`

File: test/integration/scheduler/framework_test.go [hunk 9]
```diff
@@ -1023,25 +1011,111 @@ func TestUnreservePlugin(t *testing.T) {
...
 			preBindPlugin.reset()
 			testutils.CleanupPods(testCtx.ClientSet, t, []*v1.Pod{pod})
 		})
 	}
 }
 
+// TestReservePluginOrdering tests that invocations of the unreserve
+// operation in reserve plugins is in the reverse order of invocations of
+// the reverse operation.
+func TestReservePluginOrdering(t *testing.T) {
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:1036]: Can we make these checks part of the above test?

**Review by alculquicondor — COMMENTED** [review]
Avoid doing amend while review is in progress. This is to reduce review effort.

Once your PR is approved, you can squash.

File: test/integration/scheduler/framework_test.go (hunk 9 above)
Comment (adtac) [reply test/integration/scheduler/framework_test.go:1036]: Done.

**Review by adtac — COMMENTED** [review]
Pushed PATCH v2.

File: test/integration/scheduler/framework_test.go [hunk 10]
```diff
@@ -1011,111 +1028,42 @@ func TestUnreserveReservePlugin(t *testing.T) {
 				if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
 					t.Errorf("Expected a scheduling error, but didn't get it. error: %v", err)
 				}
-				if reservePlugin.numUnreserveCalled == 0 || reservePlugin.numUnreserveCalled != preBindPlugin.numPreBindCalled {
-					t.Errorf("Expected unreserve to be called %d times, was called %d times.", preBindPlugin.numPreBindCalled, reservePlugin.numUnreserveCalled)
+				for i, pl := range reservePlugins {
+					if pl.numUnreserveCalled == 0 || pl.numUnreserveCalled != preBindPlugin.numPreBindCalled {
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:1032]: `pl.numUnreserveCalled != 1`?

File: test/integration/scheduler/framework_test.go (hunk 8 above)
Comment (alculquicondor) [reply test/integration/scheduler/framework_test.go]: sg

File: test/integration/scheduler/framework_test.go (hunk 10 above)
Comment (adtac) [reply test/integration/scheduler/framework_test.go:1032]: Done.

**Comment by adtac** [comment]
@alculquicondor Everything but the flaky e2e tests should pass now.

File: test/integration/scheduler/framework_test.go [hunk 11]
```diff
@@ -1029,7 +1029,7 @@ func TestReservePluginUnreserve(t *testing.T) {
 					t.Errorf("Expected a scheduling error, but didn't get it. error: %v", err)
 				}
 				for i, pl := range reservePlugins {
-					if pl.numUnreserveCalled == 0 || pl.numUnreserveCalled != preBindPlugin.numPreBindCalled {
+					if pl.numUnreserveCalled != 1 || pl.numUnreserveCalled != preBindPlugin.numPreBindCalled {
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:1032]: If you are intending to check different things, have a conditional for each.

**Review by alculquicondor — COMMENTED** [review]
/approve
you can squash

**Comment by alculquicondor** [comment]
State the release note in present tense:

`Unreserve extension point for scheduler plugins is merged into Reserve extension point`

File: test/integration/scheduler/framework_test.go [hunk 12]
```diff
@@ -1026,31 +1026,37 @@ func TestReservePluginUnreserve(t *testing.T) {
...
 				}
 			} else {
 				if err = testutils.WaitForPodToSchedule(testCtx.ClientSet, pod); err != nil {
-					t.Errorf("Expected the pod to be scheduled. error: %v", err)
+					t.Errorf("Expected the pod to be scheduled, got an error: %v", err)
 				}
 				for i, pl := range reservePlugins {
-					if pl.numUnreserveCalled > 0 {
-						t.Errorf("Didn't expect unreserve to be called (index %d), was called %d times.", i, pl.numUnreserveCalled)
+					if preBindPlugin.numPreBindCalled != 1 {
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:1055]: this shouldn't be inside the loop

**Comment by alculquicondor** [comment]
Please squash

/label api-review

**Comment by adtac** [comment]
rebased onto latest master, squashed, updated commit message to s/Reservation/Reserve/

File: pkg/scheduler/apis/config/types.go [hunk 13]
```diff
@@ -192,7 +192,7 @@ type Plugins struct {
 	// Score is a list of plugins that should be invoked when ranking nodes that have passed the filtering phase.
 	Score *PluginSet
 
-	// Reserve is a list of plugins invoked when reserving a node to run the pod.
+	// Reserve is a list of plugins invoked when reserving and unreserving a node to run the pod.
```
Comment (ahg-g) [inline pkg/scheduler/apis/config/types.go:195]:
```suggestion
	// Reserve is a list of plugins invoked when reserving/unreserving resources after a node is assigned to a pod.
```

File: pkg/scheduler/framework/plugins/examples/stateful/stateful.go [hunk 14]
```diff
@@ -47,14 +47,25 @@ func (mp *MultipointExample) Name() string {
...
-// PreBind is the functions invoked by the framework at "prebind" extension point.
+// Unreserve is the function invoked by the framework when any error happens
+// during "reserve" extension point or later.
+func (mp *MultipointExample) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
+	// Unlike Reserve, the Unreserve method may be called concurrently since
+	// there is no guarantee that there will only one unreserve operations at any
+	// given point in time (for example, during the binding cycle).
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+	mp.numRuns++
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/examples/stateful/stateful.go:65]: not sure this example make much sense now, and it doesn't emphasize the point that this function should be idempotent.

File: staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go [hunk 15]
```diff
@@ -175,7 +175,7 @@ type Plugins struct {
 	// Score is a list of plugins that should be invoked when ranking nodes that have passed the filtering phase.
 	Score *PluginSet `json:"score,omitempty"`
 
-	// Reserve is a list of plugins invoked when reserving a node to run the pod.
+	// Reserve is a list of plugins invoked when reserving and unreserving a node to run the pod.
```
Comment (ahg-g) [inline staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go:178]: ditto comment update.

File: pkg/scheduler/framework/plugins/examples/stateful/stateful.go (hunk 14 above)
Comment (adtac) [reply pkg/scheduler/framework/plugins/examples/stateful/stateful.go:65]: Can you elaborate? I think the example still shows plugin state can be modified at different extension points. Just like before, the Unreserve happens to be so that it may be run concurrently for different pods (just like the Prebind extension point). What has changed in terms of showcasing that feature?

Also, we do not require the Unreserve operation to be idempotent. We will never call the Unreserve function more than once for the same pod, so the authors do not need to account for that, no?

File: pkg/scheduler/scheduler.go (hunk 6 above)
Comment (mrkm4ntr) [reply pkg/scheduler/scheduler.go:527]: Oh, I misunderstood. I have nothing to do :)

File: pkg/scheduler/framework/plugins/examples/stateful/stateful.go (hunk 14 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/examples/stateful/stateful.go:65]: > Can you elaborate? I think the example still shows plugin state can be modified at different extension points.

Unreserve is supposed to reverse the effect of Reserve. Both increasing a counter in this example doesn't communicate this expectation.

>Also, we do not require the Unreserve operation to be idempotent. We will never call the Unreserve function more than once for the same pod, so the authors do not need to account for that, no?

We discussed in the past the semantics of Unreserve being idempotent, while it is true that currently we may not be calling this function more than once in the current workflow, I don't know if we can guarantee that in the future. It is easy for plugins to make this function idempotent, so I don't see the point of promising strict one-to-one call with reserve here, especially that this is an error handling extension point.

File: pkg/scheduler/apis/config/types.go (hunk 13 above)
Comment (adtac) [reply pkg/scheduler/apis/config/types.go]: Done.

File: pkg/scheduler/framework/plugins/examples/stateful/stateful.go (hunk 14 above)
Comment (adtac) [reply pkg/scheduler/framework/plugins/examples/stateful/stateful.go]: >Unreserve is supposed to reverse the effect of Reserve. Both increasing a counter in this example doesn't communicate this expectation.

Revisiting the full code once more, it looks like there are other things that could be reworkd -- I've changed the example a bit more to make it better.

re idempotency: I've added a comment in the ReservePlugin interface's Unreserve method documentation stating that the implementation must be idempotent.

File: pkg/scheduler/apis/config/types.go (hunk 13 above)
Comment (adtac) [reply pkg/scheduler/apis/config/types.go]: Done.

File: staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go (hunk 15 above)
Comment (adtac) [reply staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go]: Done.

File: test/integration/scheduler/framework_test.go [hunk 16]
```diff
@@ -1022,22 +1027,39 @@ func TestUnreservePlugin(t *testing.T) {
 
 			if test.preBindFail {
 				if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(testCtx.ClientSet, pod.Namespace, pod.Name)); err != nil {
-					t.Errorf("Expected a scheduling error, but didn't get it. error: %v", err)
+					t.Errorf("Expected a scheduling error, but didn't get it: %v", err)
 				}
-				if unreservePlugin.numUnreserveCalled == 0 || unreservePlugin.numUnreserveCalled != preBindPlugin.numPreBindCalled {
-					t.Errorf("Expected the unreserve plugin to be called %d times, was called %d times.", preBindPlugin.numPreBindCalled, unreservePlugin.numUnreserveCalled)
+				for i, pl := range reservePlugins {
+					if pl.numUnreserveCalled != 1 {
```
Comment (alculquicondor) [inline test/integration/scheduler/framework_test.go:1033]: nit: we don't need this counter, given that you are checking the channel below.

**Comment by alculquicondor** [comment]
/priority important-soon

The unit test contains a not idempotent function. But I'm ok with it, as it is a controlled environment. @ahg-g?

File: test/integration/scheduler/framework_test.go (hunk 16 above)
Comment (adtac) [reply test/integration/scheduler/framework_test.go:1033]: it still helps to check that the number of Unreserve calls isn't greater than one (the channel could have multiple buffered events from the same plugin)

Comment (adtac) [reply test/integration/scheduler/framework_test.go:1033]: actually, nvm, you're right we don't need this check given the idempotency guarantee

**Comment by adtac** [comment]
>The unit test contains a not idempotent function. But I'm ok with it, as it is a controlled environment.

@alculquicondor I think you mean the integration test (because all other Unreserve test implementations are idempotent AFAICT). And yeah, it's hard to use a idempotent function there and test everything we want to test. I've added a comment so that no one gets the wrong idea from the test.

**Comment by alculquicondor** [comment]
correct, I meant integration test :)

**Comment by ahg-g** [comment]
> /priority important-soon
> 
> The unit test contains a not idempotent function. But I'm ok with it, as it is a controlled environment. @ahg-g?

sounds good to me as long as we document the test.

**Comment by ahg-g** [comment]
/hold
/approve

pls squash

**Comment by ahg-g** [comment]
oh, we need an api review for the changes in the staging repo.

**Comment by alculquicondor** [comment]
Yes, @lavalamp is already assigned.

@adtac, please squash

File: staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go [hunk 17]
```diff
@@ -190,9 +191,6 @@ type Plugins struct {
 
 	// PostBind is a list of plugins that should be invoked after a pod is successfully bound.
 	PostBind *PluginSet `json:"postBind,omitempty"`
-
-	// Unreserve is a list of plugins invoked when a pod that was previously reserved is rejected in a later phase.
-	Unreserve *PluginSet `json:"unreserve,omitempty"`
```
Comment (lavalamp) [inline staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go:195]: Isn't v1beta1 a live type? You can't remove this field.

I think you can instead add validation that requires it to either be empty or the same as the reserve list, ignore it everywhere else, and then remove it in the next version.

**Comment by adtac** [comment]
Squashed and rebased onto latest master.

File: staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go (hunk 17 above)
Comment (alculquicondor) [reply staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go:195]: It's not. v1alpha2 was in 1.18. v1beta1 is being introduced now #89701

Comment (lavalamp) [reply staging/src/k8s.io/kube-scheduler/config/v1beta1/types.go:195]: ah ok sorry. I can't keep track of what is where.

**Comment by chendave** [comment]
/cc 

need rebase, and the cover message should be updated as well.

**Comment by adtac** [comment]
rebased onto latest master (again :))

**Comment by adtac** [comment]
The go fmt errors seem to be unrelated to my PR?

**Comment by alculquicondor** [comment]
Probably #92402 ?

**Comment by adtac** [comment]
staticcheck is failing, but gofmt is failing too and it seems like that issue doesn't cover that
